### PR TITLE
[Core] Adding ternary operator to function parser

### PR DESCRIPTION
**Pull request by loumalouomega**
**Description**
Fixes issue found after #8373. This does the ternary by manually, as tinyexpr does not support it

**Changelog**
- Added ternary operator
- Added comparison methods
- Added test

### CODE REVIEW

**Comment by philbucher** [comment]
since you already do manual parsing why not sticking with `if / else` as it was working before?

**Comment by loumalouomega** [comment]
> since you already do manual parsing why not sticking with `if / else` as it was working before?

Can you specify the notation expected?

**Comment by philbucher** [comment]
this is how it was working before:
`"modulus"         : "1.5*(0.5*(1-cos(0.5*pi*t))*2.0)*(4.0/0.1681)*y*(0.41-y) if t<2.0 else 1.5*(2.0)*(4.0/0.1681)*y*(0.41-y)",`

**Comment by philbucher** [comment]
or simplified:
`"xxx if yyy else zzz"`

**Comment by loumalouomega** [comment]
> or simplified:
> `"xxx if yyy else zzz"`

OK

**Comment by philbucher** [comment]
@rubenzorrilla maybe you could also post your usecase / example ?

**Comment by rubenzorrilla** [comment]
> @rubenzorrilla maybe you could also post your usecase / example ?

[Here](https://github.com/KratosMultiphysics/Examples/blob/89eeeade655b66aa6af94807052cb4d7c8a0e247/fluid_structure_interaction/validation/fsi_mok/source/ProjectParameters.json#L102) it is.

**Comment by loumalouomega** [comment]
> > @rubenzorrilla maybe you could also post your usecase / example ?
> 
> [Here](https://github.com/KratosMultiphysics/Examples/blob/89eeeade655b66aa6af94807052cb4d7c8a0e247/fluid_structure_interaction/validation/fsi_mok/source/ProjectParameters.json#L102) it is.

Done

**Comment by rubenzorrilla** [comment]
> > > @rubenzorrilla maybe you could also post your usecase / example ?
> > 
> > 
> > [Here](https://github.com/KratosMultiphysics/Examples/blob/89eeeade655b66aa6af94807052cb4d7c8a0e247/fluid_structure_interaction/validation/fsi_mok/source/ProjectParameters.json#L102) it is.
> 
> Done

I tried the branch with this same case and I get this error ```RuntimeError: Error: Parsing error in function: (1.0)*(0.1214*(1-cos(0.1*pi*t))*y*(1-y)```.

**Comment by loumalouomega** [comment]
> > > > @rubenzorrilla maybe you could also post your usecase / example ?
> > > 
> > > 
> > > [Here](https://github.com/KratosMultiphysics/Examples/blob/89eeeade655b66aa6af94807052cb4d7c8a0e247/fluid_structure_interaction/validation/fsi_mok/source/ProjectParameters.json#L102) it is.
> > 
> > 
> > Done
> 
> I tried the branch with this same case and I get this error `RuntimeError: Error: Parsing error in function: (1.0)*(0.1214*(1-cos(0.1*pi*t))*y*(1-y)`.

That expresion does not hold ternary expression, should work in current master

**Comment by rubenzorrilla** [comment]
> > > > > @rubenzorrilla maybe you could also post your usecase / example ?
> > > > 
> > > > 
> > > > [Here](https://github.com/KratosMultiphysics/Examples/blob/89eeeade655b66aa6af94807052cb4d7c8a0e247/fluid_structure_interaction/validation/fsi_mok/source/ProjectParameters.json#L102) it is.
> > > 
> > > 
> > > Done
> > 
> > 
> > I tried the branch with this same case and I get this error `RuntimeError: Error: Parsing error in function: (1.0)*(0.1214*(1-cos(0.1*pi*t))*y*(1-y)`.
> 
> That expresion does not hold ternary expression, should work in current master

It is the subexpression from the ternary example above. I've just copied and pasted the error message thrown.

**Comment by philbucher** [comment]
@loumalouomega I think this needs a lot more testing
=> tests for `if else` are missing
what about nesting them?

(btw I was on vacation and started writing my thesis so I am not that active atm)

**Comment by loumalouomega** [comment]
> @loumalouomega I think this needs a lot more testing
> => tests for `if else` are missing
> what about nesting them?
> 
> (btw I was on vacation and started writing my thesis so I am not that active atm)

I can do the nesting stuff, but do we really want that?

**Comment by philbucher** [comment]
> 
> 
> > @loumalouomega I think this needs a lot more testing
> > => tests for `if else` are missing
> > what about nesting them?
> > (btw I was on vacation and started writing my thesis so I am not that active atm)
> 
> I can do the nesting stuff, but do we really want that?

not sure, but at least it should throw a proper error, hence adding a test would be nice
this way it would be easier to debug and users could request it as a feature if they need it

**Comment by loumalouomega** [comment]
> > > @loumalouomega I think this needs a lot more testing
> > > => tests for `if else` are missing
> > > what about nesting them?
> > > (btw I was on vacation and started writing my thesis so I am not that active atm)
> > 
> > 
> > I can do the nesting stuff, but do we really want that?
> 
> not sure, but at least it should throw a proper error, hence adding a test would be nice
> this way it would be easier to debug and users could request it as a feature if they need it

OK, I will a error message

File: kratos/utilities/function_parser_utility.cpp [hunk 1]
```diff
@@ -141,8 +153,74 @@ void BasicGenericFunctionUtility::InitializeParser()
...
+            mpTinyExpr[0] = te_compile(mFunctionBody.c_str(), vars, 7, &err);
+            KRATOS_ERROR_IF_NOT(mpTinyExpr[0]) << "Parsing error in function: " << mFunctionBody << std::endl;
+        } else { // Ternary operator
+            mpTinyExpr.resize(3, nullptr);
+            std::string condition, first_function, second_function;
+
+            // C like ternary operator
+            std::vector<std::string> splitted_string;
+            if (c_like) {
+                splitted_string = StringUtilities::SplitStringByDelimiter(mFunctionBody, '?');
```
Comment (loumalouomega) [inline kratos/utilities/function_parser_utility.cpp]:
```suggestion
                splitted_string = StringUtilities::SplitStringByDelimiter(mFunctionBody, '?');
                KRATOS_ERROR_I(splitted_string.size() > 2) << "Nested ternary functions not supported" << std::endl;
```

File: kratos/utilities/function_parser_utility.cpp [hunk 2]
```diff
@@ -141,8 +153,74 @@ void BasicGenericFunctionUtility::InitializeParser()
...
+                condition = splitted_string[0];
+                splitted_string = StringUtilities::SplitStringByDelimiter(splitted_string[1], ':');
+                first_function = splitted_string[0];
+                second_function = splitted_string[1];
+            } else { // Python like ternary operator
+                std::string aux_string = StringUtilities::ReplaceAllSubstrings(mFunctionBody, "if", "$");
```
Comment (loumalouomega) [inline kratos/utilities/function_parser_utility.cpp]:
```suggestion
                std::string aux_string = StringUtilities::ReplaceAllSubstrings(mFunctionBody, "if", "$");
                KRATOS_ERROR_I(splitted_string.size() > 2) << "Nested ternary functions not supported" << std::endl;
```

**Comment by loumalouomega** [comment]
> > > > @loumalouomega I think this needs a lot more testing
> > > > => tests for `if else` are missing
> > > > what about nesting them?
> > > > (btw I was on vacation and started writing my thesis so I am not that active atm)
> > > 
> > > 
> > > I can do the nesting stuff, but do we really want that?
> > 
> > 
> > not sure, but at least it should throw a proper error, hence adding a test would be nice
> > this way it would be easier to debug and users could request it as a feature if they need it
> 
> OK, I will an error message

Done

File: kratos/utilities/function_parser_utility.cpp [hunk 3]
```diff
@@ -141,8 +153,76 @@ void BasicGenericFunctionUtility::InitializeParser()
...
+                KRATOS_ERROR_IF(splitted_string.size() > 2) << "Nested ternary functions not supported" << std::endl;
+                splitted_string = StringUtilities::SplitStringByDelimiter(aux_string, '$');
+                first_function = splitted_string[0];
+                aux_string = StringUtilities::ReplaceAllSubstrings(splitted_string[1], "else", "$");
+                splitted_string = StringUtilities::SplitStringByDelimiter(aux_string, '$');
+                condition = splitted_string[0];
+                second_function = splitted_string[1];
```
Comment (rubenzorrilla) [inline kratos/utilities/function_parser_utility.cpp]: We (@KratosMultiphysics/technical-committee) mean that in here you assume that the "else" is always present.

**Comment by RiccardoRossi** [comment]
what would happen if we have an "if" without the "else". This is perfectly plausible using the old interace...

example:

     if(x > 0)  x**2+y**2


as an aside **we need to test that a proper error is thrown**, which as far as i can tell is not done

**Comment by loumalouomega** [comment]
> what would happen if we have an "if" without the "else". This is perfectly plausible using the old interace...
> 
> example:
> 
> ```
>  if(x > 0)  x**2+y**2
> ```
> 
> as an aside **we need to test that a proper error is thrown**, which as far as i can tell is not done

Good point

**Comment by RiccardoRossi** [comment]
well, if we need to oblige that the else is provided, we can also live with that, as long as the error is meaningful.

actually in the error it would be good to print both the string previous to preprocessing and the string after it has been preprocessed

File: kratos/utilities/function_parser_utility.cpp (hunk 3 above)
Comment (loumalouomega) [reply kratos/utilities/function_parser_utility.cpp]: Done

**Comment by loumalouomega** [comment]
> well, if we need to oblige that the else is provided, we can also live with that, as long as the error is meaningful.
> 
> actually in the error it would be good to print both the string previous to preprocessing and the string after it has been preprocessed

Done

File: kratos/utilities/function_parser_utility.cpp [hunk 4]
```diff
@@ -141,8 +153,78 @@ void BasicGenericFunctionUtility::InitializeParser()
         const te_variable vars[] = {{"x", &x}, {"y", &y}, {"z", &z}, {"t", &t}, {"X", &X}, {"Y", &Y}, {"Z", &Z}};
 
         /* Compile the expression with variables. */
-        mpTinyExpr = te_compile(mFunctionBody.c_str(), vars, 7, &err);
-        KRATOS_ERROR_IF_NOT(mpTinyExpr) << "Parsing error in function: " << mFunctionBody << std::endl;
+        const bool c_like = StringUtilities::ContainsPartialString(mFunctionBody, "?") ? true : false;
```
Comment (philbucher) [inline kratos/utilities/function_parser_utility.cpp]: do we want to add support for this?

I feel like this is not necessary and with all the manual parsing it probably does more harm in the long run. So far the python syntax was sufficient, I don't think many ppl would use this

File: kratos/utilities/function_parser_utility.cpp [hunk 5]
```diff
@@ -141,8 +153,78 @@ void BasicGenericFunctionUtility::InitializeParser()
...
+            if (c_like) {
+                KRATOS_ERROR_IF_NOT(StringUtilities::ContainsPartialString(mFunctionBody, ":")) << "Parsing error in function: " << mFunctionBody << " ? defined, but not :" << std::endl;
+                splitted_string = StringUtilities::SplitStringByDelimiter(mFunctionBody, '?');
+                KRATOS_ERROR_IF(splitted_string.size() > 2) << "Nested ternary functions not supported" << std::endl;
+                condition = splitted_string[0];
+                splitted_string = StringUtilities::SplitStringByDelimiter(splitted_string[1], ':');
+                first_function = splitted_string[0];
+                second_function = splitted_string[1];
+            } else { // Python like ternary operator
+                KRATOS_ERROR_IF_NOT(StringUtilities::ContainsPartialString(mFunctionBody, "else")) << "Parsing error in function: " << mFunctionBody << " if defined, but not else" << std::endl;
```
Comment (philbucher) [inline kratos/utilities/function_parser_utility.cpp]: please add tests for this

**Review by philbucher — COMMENTED** [review]
some comments
I think don't need to add the ternary operator:
- natively supported 
- hence requires to change the `tinyexpr` which will make updates a huge pain
- the python works just as fine
- not sure if ppl would use the ternary operator over `if-else`

**Comment by loumalouomega** [comment]
I added this because I was asked for, and there are already tests for it.

File: kratos/utilities/function_parser_utility.cpp (hunk 4 above)
Comment (loumalouomega) [reply kratos/utilities/function_parser_utility.cpp]: No C like?

Comment (philbucher) [reply kratos/utilities/function_parser_utility.cpp]: As I said I would remove it

**Comment by loumalouomega** [comment]
Done @philbucher

File: kratos/utilities/function_parser_utility.cpp [hunk 6]
```diff
@@ -141,8 +153,67 @@ void BasicGenericFunctionUtility::InitializeParser()
...
+            mpTinyExpr[0] = te_compile(mFunctionBody.c_str(), vars, 7, &err);
+            KRATOS_ERROR_IF_NOT(mpTinyExpr[0]) << "Parsing error in function: " << mFunctionBody << std::endl;
+        } else { // Ternary operator
+            mpTinyExpr.resize(3, nullptr);
+            std::string condition, first_function, second_function;
+
+            // C like ternary operator
+            std::vector<std::string> splitted_string;
+            KRATOS_ERROR_IF_NOT(StringUtilities::ContainsPartialString(mFunctionBody, "else")) << "Parsing error in function: " << mFunctionBody << " if defined, but not else" << std::endl;
+            std::string aux_string = StringUtilities::ReplaceAllSubstrings(mFunctionBody, "if", "$");
```
Comment (philbucher) [inline kratos/utilities/function_parser_utility.cpp:167]: why is this necessary?

**Review by philbucher — COMMENTED** [review]
minor comment otherwise to me is ready

nice job :+1: 

Do you plan to push the modifications in the `tinyexpr` upstream?

File: kratos/utilities/function_parser_utility.cpp (hunk 6 above)
Comment (loumalouomega) [reply kratos/utilities/function_parser_utility.cpp:167]: Tinyexpr have some issues :P , that's why is auxiliar

**Comment by loumalouomega** [comment]
> Do you plan to push the modifications in the `tinyexpr` upstream?

Do you mean the Greater,Lesser,... functions?

**Comment by philbucher** [comment]
> > Do you plan to push the modifications in the `tinyexpr` upstream?
> 
> Do you mean the Greater,Lesser,... functions?

yes

File: kratos/utilities/function_parser_utility.cpp (hunk 6 above)
Comment (philbucher) [reply kratos/utilities/function_parser_utility.cpp:167]: can you briefly explain please, for the record?

File: kratos/tests/test_function_parser_utility.py [hunk 7]
```diff
@@ -127,6 +127,48 @@ def test_ApplyFunctionToNodesUtility(self):
...
+            utility.ApplyFunction(KM.VISCOSITY, time)
+
+            if time < 2.0:
+                for node in model_part.Nodes:
+                    self.assertEqual(node.GetSolutionStepValue(KM.VISCOSITY) - (1.5*(0.5*(1-math.cos(0.5*math.pi*time))*2.0)*(4.0/0.1681)*node.Y*(0.41-node.Y)), 0.0)
+            else:
+                for node in model_part.Nodes:
+                    self.assertAlmostEqual(node.GetSolutionStepValue(KM.VISCOSITY), 1.5*(2.0)*(4.0/0.1681)*node.Y*(0.41-node.Y))
+
+    @KratosUnittest.expectedFailure
```
Comment (philbucher) [inline kratos/tests/test_function_parser_utility.py]: please check for the actual error with `assertRaisesRegex`

**Review by philbucher — COMMENTED** [review]
@KratosMultiphysics/technical-committee to me this is ready (after the minor modification in the test)

File: kratos/tests/test_function_parser_utility.py (hunk 7 above)
Comment (loumalouomega) [reply kratos/tests/test_function_parser_utility.py]: Should be fine now

Comment (philbucher) [reply kratos/tests/test_function_parser_utility.py]:
```suggestion
```

File: kratos/tests/test_function_parser_utility.py [hunk 8]
```diff
@@ -127,6 +127,48 @@ def test_ApplyFunctionToNodesUtility(self):
...
+
+    def test_ApplyFunctionToNodesUtilityTimeEvolutionCTernaryFail(self):
+        parameters = KM.Parameters ("""{
+            "origin" : [0,0,0],
+            "axes"   : [[0,1,0],[1,0,0],[0,0,1]]
+
+        }""")
+
+        with self.assertRaisesRegex(Exception, 'Parsing error in function: 1.5 if t<2.0 3.0 if defined, but not else'):
+            function = KM.GenericFunctionUtility("1.5 if t<2.0 3.0", parameters)
```
Comment (philbucher) [inline kratos/tests/test_function_parser_utility.py]:
```suggestion
            KM.GenericFunctionUtility("1.5 if t<2.0 3.0", parameters)
```

**Comment by loumalouomega** [comment]
Ping @philbucher

**Comment by loumalouomega** [comment]
Ping @rubenzorrilla

**Review by philbucher — APPROVED** [review]
I am approving under the condition that the nightly tests pass.
I just triggered them: https://github.com/KratosMultiphysics/Kratos/actions/runs/965128347

feel free to merge if the tests pass

thx for the work!
